ssz/src/main/java/org/ethereum/beacon/ssz/SSZCodecHasher.java

```diff
@@ -0,0 +1,268 @@
+package org.ethereum.beacon.ssz;
+
+import net.consensys.cava.bytes.Bytes;
+import net.consensys.cava.ssz.BytesSSZReaderProxy;
+import net.consensys.cava.ssz.SSZ;
+import net.consensys.cava.ssz.SSZException;
+import org.ethereum.beacon.ssz.type.SSZCodec;
+import org.javatuples.Pair;
+import org.javatuples.Triplet;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * <p>Implementation of {@link SSZCodecResolver} which implements SSZ Hash function</p>
+ * <p>For more info check <a href="https://github.com/ethereum/eth2.0-specs/blob/master/specs/simple-serialize.md#tree-hash">SSZ Tree hash</a></p>
+ */
+public class SSZCodecHasher implements SSZCodecResolver {
```

Comment:
I don't think that this is a right place for hashing stuff. I mean that resolver doesn't meant to be a good place for extensions like this one. It would be great to have a codec pipeline in `SSZSerializer` that could be extended with any class that for example implements `SSZCodec`. Even `SSZCodecResolver` could be aggregated by some `PrimitiveSSZCodec implements SSZCodec`. Then to instantiating a `Hasher` would only require one more call to `SSZSerializerBuilder` comparing with regular `SSZSerializer` instantiation.

But this is just an idea. Need to check how it's applicable to the implementation we have.

Reply:
> It would be great to have a codec pipeline in SSZSerializer that could be extended with any class that for example implements SSZCodec

Currently, for example, calculating hash for container and list differs like Africa and Europe. You are trying to find object-oriented way which will combine just 2 tasks, and any of them could change dramatically tomorrow. Current solution is a simple way to deliver it fast, and any changes to hash specs could make both mine and your solutions impossible to work without hacks. I'd stay with simple and fast thing here.